Explicitly check mcause in trap handler

WasabiFan opened the pull request:
This PR introduces a distinction between exceptions we don't know how to handle and illegal instructions we don't know how to emulate. `mcause` is now explicitly checked by the handler. This should have no functional effect other than clarity.

I'd also be interested in replacing the two `while (1);` branches with something that can report the error. If there's a suggestion for how that would make sense I'd be happy to implement it.

WasabiFan committed:
Explicitly check mcause in trap handler

dpetrisko commented:
This is a pretty high overhead solution, two additional function calls. I would take a look at how riscv-pk / opensbi handle this (function pointer tables)

WasabiFan commented:
That's fair. I'd hope that the compiler with optimizations enabled would inline the functions and produce a jump table if the switch got big enough, but at least the former doesn't appear to be the case automatically.

Regarding looking into what others do -- for the top-level exception handler delegation, [opensbi uses a switch](https://github.com/riscv/opensbi/blob/49e422c5ad7abee9be0bb1066b1db65676e61985/lib/sbi/sbi_trap.c#L242) while [riscv-pk uses a jump table](https://github.com/riscv/riscv-pk/blob/66d7fcb56d6a4cd4879922f184bb2274918ac3cd/machine/mentry.S#L9). For identifying whether a particular opcode is emulate-able, both [opensbi](https://github.com/riscv/opensbi/blob/49e422c5ad7abee9be0bb1066b1db65676e61985/lib/sbi/sbi_illegal_insn.c#L82) and [riscv-pk](https://github.com/riscv/riscv-pk/blob/66d7fcb56d6a4cd4879922f184bb2274918ac3cd/machine/emulation.c#L71) use jump tables.

I'll take another look at this when I get the chance. Minimal function calls plus a jump table for one or both of the above are absolutely options. I think persuading the compiler to inline the instruction emulations, or making that its own jump-table at the top level to avoid the calls altogether, might be good ways to go.

farzamgl commented:
Great PR. I just want to add that it's preferable to terminate the program(with an error msg) on an unhandled exception instead of looping. You can also check mcause in the trap handler assembly, so the common overhead would be 1 extra branch op.

WasabiFan commented:
I'm happy to do that. Would terminating with an error message mean a series of `bp_cprint` and then a `bp_finish(1)`? I haven't seen existing infrastructure for things like this, so perhaps I'll throw in a `panic(message)` or similar for general use if that seems helpful.

dpetrisko commented:
In the bp_nonsynth_host I believe we print FAIL if we receive a finish packet with a nonzero error code

dpetrisko commented:
A panic function that takes a string, prints it and terminates with error would be a useful addition

WasabiFan commented:
I've updated the PR; I think this now incorporates the above feedback. I implemented the mcause-based delegation logic directly in the asm trap handler so it should now be fairly lightweight.

If many more handlers are added in the future, the comparisons should probably become a jump table, but for now I kept it simple to avoid having to do bounds checks. Happy to change if preferred.

What are your thoughts on #1? If that is also worth merging, I may want to adopt the `bp_hprint_uint64` for dumping context (mcause, mtval) in the case of an unhandled trap. Alternately I could just special-case the string message for common mcause (illegal accesses, unaligned accesses, etc.).

dpetrisko commented:
I would throw the "unhandled_trap" code into emulation.c/h.

Otherwise looks good! I merged #1 so rebase to fix that conflict, then I think we're good

WasabiFan commented:
Updated. I have moved the hander to `emulation.c` and introduced the following into the trap abort handler:

```c
  bp_print_string("mcause: ");
  bp_hprint_uint64(mcause);
  bp_print_string("\nmtval: ");
  bp_hprint_uint64(mtval);
```

This looks as follows:

```
Hello World!
mcause: 0x0000000000000003
mtval: 0x0000000000000000
Unhandled machine-mode trap
[CORE0 FSH] FAIL
```

dpetrisko commented:
Awesome! Could you add a test which traps to an unhandled exception to bp-tests?

farzamgl commented:
> Updated. I have moved the hander to `emulation.c` and introduced the following into the trap abort handler:
> 
> ```c
>   bp_print_string("mcause: ");
>   bp_hprint_uint64(mcause);
>   bp_print_string("\nmtval: ");
>   bp_hprint_uint64(mtval);
> ```
> 
> This looks as follows:
> 
> ```
> Hello World!
> mcause: 0x0000000000000003
> mtval: 0x0000000000000000
> Unhandled machine-mode trap
> [CORE0 FSH] FAIL
> ```

farzamgl commented:
Looks good! Can you also add the mepc value to the unhandled exception msg? Will help a lot with debugging.

WasabiFan commented:
Good call! Updated:

```
Hello World!
mcause: 0x0000000000000003
mtval:  0x0000000000000000
mepc:   0x0000000080000c58
Unhandled machine-mode trap
[CORE0 FSH] FAIL
```

I'll open a PR against bp-tests later today.

WasabiFan commented:
Just threw in a pc readout in the illegal instruction abort too.

```
Hello World!
instr addr: 0x00000000800eefe8
Illegal instruction
[CORE0 FSH] FAIL
```

farzamgl commented:
You can just call the unhandled_trap_abort in the illegal case to unify the error format.

WasabiFan commented:
Is that preferable over having the separate error strings? It just saves you from looking up the mcause values and behavior of mepc.

farzamgl commented:
I think if you get to that point and want to debug the program you need to lookup that stuff, but either way is fine with me.

WasabiFan commented:
Here's some cleanup 🙂 Now it provides full context for both, with separate messages.

```diff
diff --git a/emulation.c b/emulation.c
index beffb7a..3abacce 100644
--- a/emulation.c
+++ b/emulation.c
@@ -46,6 +46,19 @@ static uint64_t (*mul_jt[8])(uint64_t, uint64_t) =
   0, mul_mulh, mul_mulhsu, mul_mulhu, 0, 0, 0, 0
 };
 
+void dump_trap_context(uint64_t mcause, uint64_t mtval)
+{
+  uint64_t mepc;
+  __asm__ __volatile__ ("csrr %0, mepc" : "=r" (mepc));
+
+  bp_print_string("mcause: ");
+  bp_hprint_uint64(mcause);
+  bp_print_string("\nmtval:  ");
+  bp_hprint_uint64(mtval);
+  bp_print_string("\nmepc:   ");
+  bp_hprint_uint64(mepc);
+}
+
 void emulate_illegal_instruction(uint64_t *regs, uint64_t mcause, uint64_t instr)
 {
   uint16_t funct11 = FUNCT11(instr);
@@ -66,11 +79,7 @@ void emulate_illegal_instruction(uint64_t *regs, uint64_t mcause, uint64_t instr
     regs[rd_addr] = mul_jt[funct3](rs1_data, rs2_data);
   } else {
     // Fail on truly illegal instruction
-    uint64_t mepc;
-    __asm__ __volatile__ ("csrr %0, mepc" : "=r" (mepc));
-
-    bp_print_string("instr addr: ");
-    bp_hprint_uint64(mepc);
+    dump_trap_context(mcause, instr);
     bp_panic("\nIllegal instruction\n");
   }
 }
@@ -79,11 +88,6 @@ void unhandled_trap_abort(uint64_t *regs, uint64_t mcause, uint64_t mtval) {
   uint64_t mepc;
   __asm__ __volatile__ ("csrr %0, mepc" : "=r" (mepc));
 
-  bp_print_string("mcause: ");
-  bp_hprint_uint64(mcause);
-  bp_print_string("\nmtval:  ");
-  bp_hprint_uint64(mtval);
-  bp_print_string("\nmepc:   ");
-  bp_hprint_uint64(mepc);
+  dump_trap_context(mcause, mtval);
   bp_panic("\nUnhandled machine-mode trap\n");
 }

```

WasabiFan commented:
Is this good to merge to go along with black-parrot-sdk/bp-tests#10?

dpetrisko reviewed (approved):
LGTM

dpetrisko commented:
@farzamgl look good?

farzamgl commented on emulation.c:
```diff
@@ -66,8 +79,15 @@ void decode_illegal(uint64_t *regs, uint64_t mcause, uint64_t instr)
 }
 
+void unhandled_trap_abort(uint64_t *regs, uint64_t mcause, uint64_t mtval) {
+  uint64_t mepc;
```
This line is extra I think.

farzamgl reviewed (commented):
I think it's wise to add a "bp_" prefix to the trap handling functions' names to avoid conflicting symbol names when building a test with libperch.

WasabiFan commented:
Updated to remove dead code and add `bp_` to all fn names.

farzamgl commented on exception.S:
```diff
@@ -4,8 +4,10 @@
 
-.globl decode_illegal
-.type  decode_illegal, @function
+.globl handle_exception
```
This also looks extra.

WasabiFan replied:
Do you know why it was there originally? To my knowledge neither the original or new form of this would _no anything_, but I kept it assuming there was some intent I didn't understand. Should I just remove it?

farzamgl replied:
I thought you added it, didn't you?

farzamgl replied:
Ah I see what you mean. Yeah I don't think the .globl did anything here before, but also `handle_exception` doesn't exist so there's no reason to have it now.

WasabiFan replied:
Good points on both; removed.

farzamgl reviewed (approved):
One minor comment. Other than that looks great!

WasabiFan commented:
I just rebased this on latest master, but it should still be good-to-go.

dpetrisko committed:
Merge branch 'master' into trap-mcause
